### PR TITLE
PSR2 MethodArgumentSpacerFixer clean additional spaces

### PR DESCRIPTION
**Pull request by trilopin**
PSR2 MethodArgumentSpacerFixer clean additional spaces after a comma.
With this code one and only one space can be after a comma in a method call.

> 4.4. Method Arguments
> In the argument list, there MUST NOT be a space before each comma, and there MUST be one space after each comma.

Thanks

### CODE REVIEW

**Comment by trilopin** [comment]
Hi, i have pushed the code for the last testcase (plus another more complex).
php-cs-fixer and phpunit say that all are running locally without errors but not at travis.

Any idea?

Thanks

**Comment by keradus** [comment]
Please rebase your local work on upstream/master

**Comment by GrahamCampbell** [comment]
master?

**Comment by keradus** [comment]
> 	trilopin  wants to merge 26 commits into FriendsOfPHP:master from trilopin:master

So Travis apply patch from this PR into `master` branch.
@trilopin run it locally and has tests passed, so I guess that he doesn't work on latest version of branch

**Comment by trilopin** [comment]
After rebase I see two changes, the constant for block braces and newline now has in his own token. 
i have changed the code and now is running ok again.

Sorry, I think that i never should have done pr from master..

Is safe to push now after rebase from upstream?

**Comment by keradus** [comment]
I do it often... Just remember to use `-f` switch (force)

File: Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php [hunk 1]
```diff
@@ -29,27 +29,43 @@ public function fix(\SplFileInfo $file, $content)
     {
         $tokens = Tokens::fromCode($content);
 
-        for ($index = $tokens->count() - 1; $index >= 0; --$index) {
+        for ($index = $tokens->count() - 1; 0 <= $index; --$index) {
```
Comment (stof) [inline Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php]: why changing this ?

File: Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php [hunk 2]
```diff
@@ -69,13 +85,46 @@ public function fixSpace(Tokens $tokens, $index)
...
+     * Check if last item of current line is a comment
+     *
+     * @param Tokens $tokens tokens to handle
+     * @param int    $index  index of token
+     *
+     * @return bool
+     */
+    private function isCommentLastLineToken(Tokens $tokens, $index)
+    {
+        return $tokens[$index]->isComment() && 1 === substr_count($tokens[$index + 1]->getContent(), "\n");
```
Comment (stof) [inline Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php]: Will this work properly for multiline comments ? ``"/** foo\nbar*/"`` ?

Comment (GrahamCampbell) [inline Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php]: Multiline comments are not split up. They are all one token `T_DOC_COMMENT`.

Comment (GrahamCampbell) [inline Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php]: So, yeh, it will work just the same.

Comment (stof) [inline Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php]: I edited my comment to fix a typo in it. I'm talking about multiline comments, **not about T_DOC_COMMENTS**. Not all multiline comments are doc comments. Doc comments are starting with ``/**``

Comment (keradus) [inline Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php]:
```php
/*
 * foo
bar
*/
```

 is still one `T_COMMENT`, isn't it?

Comment (GrahamCampbell) [inline Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php]: yeh, it is i think

Comment (GrahamCampbell) [inline Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php]: yeh, it must be given that behaviour with the new lines before.

**Comment by trilopin** [comment]
I just tested with following cases with sucess (without code changes):

```php
            // must keep align comments
            array(
                '<?php function xyz(
                    $a=10,      /* comment1 */
                    $b=20,      /* comment2     */
                    $c=30) {
                }',
            ),
            array(
                '<?php function xyz(
                    $a=10,  /*comment1*/
                    $b=2000,/*comment2*/
                    $c=30) {
                }',
            ),
```

Are this comments that you say? do you want to add this testcases to PR?

File: Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php (hunk 1 above)
Comment (trilopin) [inline Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php]: i took the solution that @keradus said me from strcitparamsfixer. 

There is no specific reason for one or another version,

**Comment by keradus** [comment]
> Will this work properly for multiline comments ? "/* foo\nbar*/" ?

He was talking about **multiline** comments ;)

**Comment by keradus** [comment]
OK, so technically we can't merge that branch currently.
Please do the following:
- squash consecutive commits of one author
- rebase branch on upstream/master

Feel free to ask for help on it, if you need any

**Comment by trilopin** [comment]
I know how to rebase interactive for squash commits but I never have done
in pushed ones.

So, will be safe this flow? Is needed push forced?

Thanks

**Comment by stof** [comment]
@trilopin you will need to force the push to your branch after rebasing, yes. For branches used to send a PR, we consider it is OK because the commits are not yet part of the git history of the library.

**Comment by trilopin** [comment]
ok, pr has been rebased from upstream and commits joined

**Comment by keradus** [comment]
great ;)


:+1:

**Comment by keradus** [comment]
Then one last thing before merge. Could you deal with https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/894#discussion_r22423031 @trilopin  ?

**Comment by trilopin** [comment]
I have added testcase for multline comments:

```php
            //multiline comments also must be ignored
            array(
                '<?php function xyz(
                    $a=10,  /* comment1a
                               comment1b
                            */
                    $b=2000,/* comment2a
                        comment 2b
                        comment 2c */
                    $c=30) {
                }',
            ),
```

**Comment by trilopin** [comment]
@keradus few days ago I added testcase for multiline comments

is anything needed to merge pr?

**Comment by keradus** [comment]
I just wait for @stof to answer.
Very well.
:+1:

I will merge it this evening

**Comment by stof** [comment]
@trilopin the case which would break with your current code is not the one you suggested in your comment. It is the case where you have have a multiline command and the code on the same line than the end of the comment (i.e. the multiline comment is not the last token on the line where it ends). Your ``isCommentLastLineToken`` currently returns true for this case

**Comment by trilopin** [comment]
@stof, could you provide code for this testcase? 

I still dont understand which case you refer to.

Thanks

**Comment by stof** [comment]
```php
//multiline comments also must be ignored
            array(
                '<?php function xyz(
                    $a=10, /* multiline comment
                             not at the end of line
                            */ $b=2000,
                    $c=30) {
                }',
            ),
```

Your method ``isCommentLastLineToken`` returns true for this comment, which is wrong.

**Comment by keradus** [comment]
@stof you want that case to be changed into sth or what?
After adding it to tests as is - it passes.

**Comment by keradus** [comment]
Please see https://github.com/trilopin/PHP-CS-Fixer/pull/3
// cc @trilopin @stof

**Comment by trilopin** [comment]
@keradus Last week  I've tried to rebase from upstream/master with weird results...  After rebase and resolve conflics, some code is a mess.  In fact I coded a solution for multiline comment case but i did not push because of difficult rebase. 

I am in same case with your PR. 

Could you help with rebase? Could be better to do a clean PR from current master?

Thanks

**Comment by keradus** [comment]
Ow.. You have did your changes on master branch. When you work on forked repo it's cool to not work on original branches but create your own.
It will be a little tricky now, but it can be done ;)

The following part will squash commits. My first one is needed for yours second one, therefore it can be ommited in commands.
```bash
git co 27b537907b269194bcd3dbf18006a6852c1efab4
git cherry-pick c091b5021b5a8d11e762d782cf8223a06a6b1fd7 --no-commit
git checkout --theirs .
git add .
git commit --amend -m "MethodArgumentSpaceFixer - enhance sth..."
git cherry-pick 803a26182d3754509b1efddc4701ba8d88114465
git br -D master
git co -b master
git diff origin/master HEAD ## check if you should see no changes here - to be sure everything is ok, then...
git poc -f
```

Now you need to rebase on master, starting by
```bash
git fetch upstream
git rebase upstream/master
```


Then you will need to manually change your code to face new upstream - on upstream/master we changed methods order and interface of `fix` method.

**Comment by trilopin** [comment]
I have done (resolving aliases):
```
git co 27b537907b269194bcd3dbf18006a6852c1efab4
git cherry-pick c091b5021b5a8d11e762d782cf8223a06a6b1fd7 --no-commit
git checkout --theirs .
git add .
git commit --amend -m "MethodArgumentSpaceFixer - enhance sth..."
git cherry-pick 803a26182d3754509b1efddc4701ba8d88114465
git branch -D master
git checkout -b master
git diff origin/master HEAD 
git push -u -f origin master 
```

Now, i am doing

```
git fetch upstream
git rebase upstream/master
First, rewinding head to replay your work on top of it...
Applying: MethodArgumentSpaceFixer - enhance fix, only one space after comma
Using index info to reconstruct a base tree...
M	Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php
Falling back to patching base and 3-way merge...
Automezclado Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php
CONFLICTO(contenido): conflicto de fusión en Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php
Failed to merge in the changes.
Patch failed at 0001 MethodArgumentSpaceFixer - enhance fix, only one space after comma
The copy of the patch that failed is found in:
   /home/jpeso/git-projects/PHP-CS-Fixer/.git/rebase-apply/patch

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
```

resolv conflict
```
git add Symfony/CS/Fixer/PSR2/MethodArgumentSpaceFixer.php
git rebase --continue
```

code is messed after resolving commit in methodargumentfixer, for example if i go to fixerinterface i see the old code.
```php
public function fix(\SplFileInfo $file, Tokens $tokens);
```
 since last week happens the same thing when rebase.

**Comment by keradus** [comment]
Yeah, the whole files is conflicted due to 2 things.
> Then you will need to manually change your code to face new upstream - on upstream/master
> we changed methods order and interface of fix method.

You must manually resolve whole files, not only content between `<<<<<<` etc.

**Comment by GrahamCampbell** [comment]
This would be a lot less confusing if this wasn't targeted against the master branch on github....

**Comment by keradus** [comment]
It uses new transformer from master and cannot be target against 1.x

**Comment by GrahamCampbell** [comment]
> It uses new transformer from master and cannot be target against 1.x

Sorry, forgot about that. :)

**Comment by keradus** [comment]
Thank you @trilopin.

Merged manually in #998

**Comment by trilopin** [comment]
Thanks, merge was too complicated for me ;)

**Comment by keradus** [comment]
No problem, I'm glad I could help ;)

**Pull request closed without merging**